In [1]:
# dataset info
import pandas as pd

df = pd.read_excel(
    "../data/online_retail_II xlsx kaggle.xlsx"
)

df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [2]:
# missing values
df.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

In [3]:
# duplicate record count
df.duplicated().sum()

np.int64(6865)

In [4]:
#removing duplicates
df = df.drop_duplicates()

### Duplicate Records Handling

The dataset contained 6,865 duplicate records.
These duplicates were removed to ensure accurate sales, customer, and inventory analysis.

In [5]:
df.shape

(518596, 8)

In [6]:
# missing values
df.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107833
Country             0
dtype: int64

In [7]:
# removing missing customer IDs
df = df.dropna(subset=["Customer ID"])

In [8]:
df.shape

(410763, 8)

In [9]:
# removing missing descriptions
df = df.dropna(subset=["Description"])

In [10]:
df.shape

(410763, 8)

### Handling Missing Values

- Description column contained 2,928 missing values.
- Customer ID column contained 107,833 missing values.

Rows with missing Customer IDs were removed because customer identification is essential for RFM analysis, churn prediction, and customer segmentation.

Rows with missing product descriptions were also removed to maintain data quality.

In [11]:
# negative quantities (returns/cancellations)
(df["Quantity"] < 0).sum()

np.int64(9816)

In [12]:
df = df[df["Quantity"] > 0]

In [13]:
df.shape

(400947, 8)

In [14]:
# invalid prices
(df["Price"] <= 0).sum()

np.int64(31)

In [15]:
df = df[df["Price"] > 0]

In [16]:
df.shape

(400916, 8)

In [17]:
# creating revenue
df["Revenue"] = df["Quantity"] * df["Price"]

In [18]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0


In [19]:
# converting invoice date
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [20]:
# creating date features
df["Year"] = df["InvoiceDate"].dt.year

In [21]:
df["Month"] = df["InvoiceDate"].dt.month

In [22]:
df["Day"] = df["InvoiceDate"].dt.day

In [23]:
df["Weekday"] = df["InvoiceDate"].dt.day_name()

In [24]:
df.shape

(400916, 13)

In [25]:
df.head

<bound method NDFrame.head of        Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
525456  538171     22271                 FELTCRAFT DOLL ROSIE         2   
525457  538171     22750         FELTCRAFT PRINCESS LOLA DOLL         1   
525458  538171     22751       FELTCRAFT PRINCESS OLIVIA DOLL         1   
525459  538171     20970   PINK FLORAL FELTCRAFT SHOULDER BAG         2   
525460  538171     21931               JUMBO STORAGE BAG SUKI         2   

               InvoiceDate  Price  Customer ID         Country  Reven

In [26]:
df.to_csv(
    "../data/cleaned_retail.csv",
    index=False
)